In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import pyodbc
import sqlalchemy
import sqlite3
from subprocess import check_output
import os

%sql sqlite://

'Connected: @None'

## Importing & viewing dataset

In [4]:
currencies = pd.read_csv('/kaggle/input/countriescurrencies/currencies.csv', 
                           index_col = 'code')
print(currencies.info())
currencies.head()

<class 'pandas.core.frame.DataFrame'>
Index: 224 entries, AFG to ZWE
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   curr_id        224 non-null    int64  
 1   basic_unit     224 non-null    object 
 2   curr_code      224 non-null    object 
 3   frac_unit      223 non-null    object 
 4   frac_perbasic  223 non-null    float64
dtypes: float64(1), int64(1), object(3)
memory usage: 10.5+ KB
None


curr_id      basic_unit curr_code frac_unit  frac_perbasic
code                                                            
AFG         1  Afghan afghani       AFN       Pul          100.0
ALB         2    Albanian lek       ALL  Qindarke          100.0
DZA         3  Algerian dinar       DZD   Santeem          100.0
AND         4            Euro       EUR      Cent          100.0
AGO         5  Angolan kwanza       AOA   Centimo          100.0

In [5]:
countries = pd.read_csv('/kaggle/input/countriescurrencies/countries.csv', index_col = 'code')
print(countries.info())
countries.head()

<class 'pandas.core.frame.DataFrame'>
Index: 206 entries, AFG to PSE
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   country_name  206 non-null    object 
 1   continent     206 non-null    object 
 2   region        206 non-null    object 
 3   surface_area  206 non-null    float64
 4   indep_year    188 non-null    float64
 5   local_name    206 non-null    object 
 6   gov_form      206 non-null    object 
 7   capital       201 non-null    object 
 8   cap_long      204 non-null    float64
 9   cap_lat       204 non-null    float64
dtypes: float64(4), object(6)
memory usage: 17.7+ KB
None


country_name continent                     region  surface_area  \
code                                                                      
AFG      Afghanistan      Asia  Southern and Central Asia      652090.0   
NLD      Netherlands    Europe             Western Europe       41526.0   
ALB          Albania    Europe            Southern Europe       28748.0   
DZA          Algeria    Africa            Northern Africa     2381740.0   
ASM   American Samoa   Oceania                  Polynesia         199.0   

      indep_year             local_name                 gov_form    capital  \
code                                                                          
AFG       1919.0  Afganistan/Afqanestan          Islamic Emirate      Kabul   
NLD       1581.0              Nederland  Constitutional Monarchy  Amsterdam   
ALB       1912.0              Shqiperia                 Republic     Tirane   
DZA       1962.0     Al-Jaza’ir/Algerie                 Republic    Algiers   
ASM          NaN          Amerika Samoa             US Territory  Pago Pago   

       cap_long  cap_lat  
code                      
AFG    69.17610  34.5228  
NLD     4.89095  52.3738  
ALB    19.81720  41.3317  
DZA     3.05097  36.7397  
ASM  -170.69100 -14.2846

In [6]:
cities = pd.read_csv('/kaggle/input/countriescurrencies/cities.csv', 
                          index_col = 'name')
print(cities.info())
cities.head()

<class 'pandas.core.frame.DataFrame'>
Index: 236 entries, Abidjan to Zunyi
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   country_code     236 non-null    object 
 1   city_proper_pop  236 non-null    int64  
 2   metroarea_pop    137 non-null    float64
 3   urbanarea_pop    236 non-null    int64  
dtypes: float64(1), int64(2), object(1)
memory usage: 9.2+ KB
None


country_code  city_proper_pop  metroarea_pop  urbanarea_pop
name                                                                   
Abidjan              CIV          4765000            NaN        4765000
Abu Dhabi            ARE          1145000            NaN        1145000
Abuja                NGA          1235880      6000000.0        1235880
Accra                GHA          2070463      4010054.0        2070463
Addis Ababa          ETH          3103673      4567857.0        3103673

In [7]:
economies = pd.read_csv('/kaggle/input/countriescurrencies/economies.csv', 
                           index_col = 'code')
print(economies.info())
economies.head()

<class 'pandas.core.frame.DataFrame'>
Index: 380 entries, AFG to ZWE
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   econ_id            380 non-null    int64  
 1   year               380 non-null    int64  
 2   income_group       380 non-null    object 
 3   gdp_percapita      378 non-null    float64
 4   gross_savings      338 non-null    float64
 5   inflation_rate     377 non-null    float64
 6   total_investment   344 non-null    float64
 7   unemployment_rate  217 non-null    float64
 8   exports            351 non-null    float64
 9   imports            349 non-null    float64
dtypes: float64(7), int64(2), object(1)
memory usage: 32.7+ KB
None


econ_id  year         income_group  gdp_percapita  gross_savings  \
code                                                                     
AFG         1  2010           Low income        539.667         37.133   
AFG         2  2015           Low income        615.091         21.466   
AGO         3  2010  Upper middle income       3599.270         23.534   
AGO         4  2015  Upper middle income       3876.200         -0.425   
ALB         5  2010  Upper middle income       4098.130         20.011   

      inflation_rate  total_investment  unemployment_rate  exports  imports  
code                                                                         
AFG            2.179            30.402                NaN   46.394   24.381  
AFG           -1.549            18.602                NaN  -49.110   -7.294  
AGO           14.480            14.433                NaN   -3.266  -21.076  
AGO           10.287             9.552                NaN    6.721  -21.778  
ALB            3.605            31.305               14.0   10.645   -8.013

In [8]:
economies2010 = pd.read_csv('/kaggle/input/countriescurrencies/economies2010.csv', 
                           index_col = 'code')
print(economies2010.info())
economies2010.head()

<class 'pandas.core.frame.DataFrame'>
Index: 190 entries, AFG to ZWE
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   year           190 non-null    int64  
 1   income_group   190 non-null    object 
 2   gross_savings  169 non-null    float64
dtypes: float64(1), int64(1), object(1)
memory usage: 5.9+ KB
None


year         income_group  gross_savings
code                                          
AFG   2010           Low income         37.133
AGO   2010  Upper middle income         23.534
ALB   2010  Upper middle income         20.011
ARE   2010          High income         27.073
ARG   2010  Upper middle income         17.361

In [9]:
economies2015 = pd.read_csv('/kaggle/input/countriescurrencies/economies2015.csv', 
                           index_col = 'code')
print(economies2015.info())
economies2015.head()

<class 'pandas.core.frame.DataFrame'>
Index: 190 entries, AFG to ZWE
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   year           190 non-null    int64  
 1   income_group   190 non-null    object 
 2   gross_savings  169 non-null    float64
dtypes: float64(1), int64(1), object(1)
memory usage: 5.9+ KB
None


year         income_group  gross_savings
code                                          
AFG   2015           Low income         21.466
AGO   2015  Upper middle income         -0.425
ALB   2015  Upper middle income         13.840
ARE   2015          High income         34.106
ARG   2015  Upper middle income         14.111

In [10]:
languages = pd.read_csv('/kaggle/input/countriescurrencies/languages.csv', 
                           index_col = 'code')
print(languages.info())
languages.head()

<class 'pandas.core.frame.DataFrame'>
Index: 955 entries, AFG to ZWE
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   lang_id   955 non-null    int64  
 1   name      955 non-null    object 
 2   percent   518 non-null    float64
 3   official  955 non-null    bool   
dtypes: bool(1), float64(1), int64(1), object(1)
memory usage: 30.8+ KB
None


lang_id      name  percent  official
code                                      
AFG         1      Dari     50.0      True
AFG         2    Pashto     35.0      True
AFG         3    Turkic     11.0     False
AFG         4     Other      4.0     False
ALB         5  Albanian     98.8      True

In [11]:
populations = pd.read_csv('/kaggle/input/countriescurrencies/populations.csv', 
                           index_col = 'pop_id')
print(populations.info())
populations.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 434 entries, 20 to 432
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   country_code     434 non-null    object 
 1   year             434 non-null    int64  
 2   fertility_rate   402 non-null    float64
 3   life_expectancy  398 non-null    float64
 4   size             433 non-null    float64
dtypes: float64(3), int64(1), object(1)
memory usage: 20.3+ KB
None


country_code  year  fertility_rate  life_expectancy        size
pop_id                                                                
20              ABW  2010           1.704        74.953537    101597.0
19              ABW  2015           1.647        75.573585    103889.0
2               AFG  2010           5.746        58.970829  27962207.0
1               AFG  2015           4.653        60.717171  32526562.0
12              AGO  2010           6.416        50.654171  21219954.0

## Plug-into SQL database

In [12]:
from sqlalchemy import create_engine
engine = create_engine('sqlite:////COUNTRIES_INFO_DETAILS', echo=False)
countries.to_sql('countries', con = engine)
economies.to_sql('economies', con = engine)
economies2010.to_sql('economies2010', con = engine)
economies2015.to_sql('economies2015', con = engine)
currencies.to_sql('currencies', con = engine)
cities.to_sql('cities', con = engine)
languages.to_sql('languages', con = engine)
populations.to_sql('populations', con = engine)

# PRACTICEs

## 1. Union

### 1.1. Union (basic : non-including duplicates)

Near query result to the right, you will see two new tables with names economies2010 and economies2015.

#### Instructions

Combine these two tables into one table containing all of the fields in economies2010. The economies table is also included for reference.

Sort this resulting single table by country code and then by year, both in ascending order.

In [13]:
pd.read_sql(
    """
        SELECT *
        -- From 2010 table
        FROM economies2010
        -- Set theory clause
        UNION
        -- Select fields from 2015 table
        SELECT *
          -- From 2015 table
        FROM economies2015
        -- Order by code and year
        ORDER BY code, year;
    """, con = engine)

code  year         income_group  gross_savings
0    AFG  2010           Low income         37.133
1    AFG  2015           Low income         21.466
2    AGO  2010  Upper middle income         23.534
3    AGO  2015  Upper middle income         -0.425
4    ALB  2010  Upper middle income         20.011
..   ...   ...                  ...            ...
375  ZAF  2015  Upper middle income         16.460
376  ZMB  2010  Lower middle income         37.404
377  ZMB  2015  Lower middle income         39.177
378  ZWE  2010           Low income         16.109
379  ZWE  2015           Low income          5.563

[380 rows x 4 columns]

**UNION** can also be used to determine all occurrences of a field across multiple tables. Try out this exercise with no starter code.

Instructions
100 XP
Determine all (non-duplicated) country codes in either the cities or the currencies table. The result should be a table with only one field called country_code.
Sort by country_code in alphabetical order.


In [14]:
pd.read_sql(
    """
        SELECT country_code
        FROM cities
        UNION
        SELECT code
        FROM currencies
        ORDER BY country_code;
    """, con = engine)

country_code
0            ABW
1            AFG
2            AGO
3            AIA
4            ALB
..           ...
200          WSM
201          YEM
202          ZAF
203          ZMB
204          ZWE

[205 rows x 1 columns]

### 1.2. Union all

As you saw, `duplicates` were removed from the previous two exercises by using `UNION`.

**To include duplicates**, you can use **`UNION ALL`**.

#### Instructions

Determine all combinations (include duplicates) of country code and year that exist in either the economies or the populations tables. Order by code then year.

The result of the query should only have two columns/fields. Think about how many records this query should result in.

You'll use code very similar to this in your next exercise after the video. Make note of this code after completing it

In [15]:
pd.read_sql(
    """
        SELECT code, year
        FROM economies
        UNION ALL
        SELECT country_code, year
        FROM populations
        ORDER BY code, year;
    """, con = engine)

code  year
0    ABW  2010
1    ABW  2015
2    AFG  2010
3    AFG  2010
4    AFG  2015
..   ...   ...
809  ZMB  2015
810  ZWE  2010
811  ZWE  2010
812  ZWE  2015
813  ZWE  2015

[814 rows x 2 columns]

## 2. Intersect

Repeat the previous **`UNION ALL`** exercise, this time looking at the records in common for `country_code` and `year` for the `economies` and `populations tables`.

#### Instructions

Again, order by code and then by year, both in ascending order.

Note the number of records here (given at the bottom of query result) compared to the similar UNION ALL query result (814 records)

In [16]:
pd.read_sql(
    """
        SELECT code, year
        FROM economies
        INTERSECT
        SELECT country_code, year
        FROM populations
        ORDER BY code, year;
    """, con = engine)

code  year
0    AFG  2010
1    AFG  2015
2    AGO  2010
3    AGO  2015
4    ALB  2010
..   ...   ...
375  ZAF  2015
376  ZMB  2010
377  ZMB  2015
378  ZWE  2010
379  ZWE  2015

[380 rows x 2 columns]

As you think about major world cities and their corresponding country, you may ask which countries also have a city with the same name as their country name?

#### Instruction.

Use **INTERSECT** to answer this question with countries and cities!

In [17]:
pd.read_sql(
    """
        SELECT code, country_name
        FROM countries
        INTERSECT
        SELECT country_code, name
        FROM cities
    """, con = engine)

code country_name
0  SGP    Singapore

## 3. Except
Get the names of cities in cities which are not noted as capital cities in countries as a single field result.

Note that there are some countries in the world that are not included in the countries table, which will result in some cities not being labeled as capital cities when in fact they are.


#### Instructions

Order the resulting field in ascending order.

Can you spot the city/cities that are actually capital cities which this query misses

In [18]:
pd.read_sql(
    """
        SELECT name
        FROM cities
        EXCEPT
        SELECT capital
        FROM countries
        ORDER BY name;    
    """, con = engine)

name
0          Abidjan
1        Ahmedabad
2       Alexandria
3           Almaty
4         Auckland
..             ...
165  Yekaterinburg
166       Yokohama
167      Zhengzhou
168      Zhongshan
169          Zunyi

[170 rows x 1 columns]

Remember that; the normal **exception** of any 2 set has 2 cases
$$ A \backslash B $$
and
$$ B \backslash A $$
#### Exercise 3.2.
Now you will complete the previous query in `reverse`!

Determine the names of capital cities that are not listed in the cities table. In this case; you must use `capital_countries \ name_cities`

**Instructions**

Order by capital in ascending order.
The cities table contains information about 236 of the world's most populous cities. The result of your query may surprise you in terms of the number of capital cities that DO NOT appear in this list!

In [19]:
pd.read_sql(
    """
        SELECT capital
        FROM countries
        EXCEPT
        SELECT name
        FROM cities
        ORDER BY capital;
    """, con = engine)

capital
0                None
1               Agana
2               Amman
3           Amsterdam
4    Andorra la Vella
..                ...
131        Wellington
132          Windhoek
133      Yamoussoukro
134    Yaren District
135            Zagreb

[136 rows x 1 columns]

## 4. Semi-join.

You are now going to use the concept of a semi-join to identify languages spoken in the Middle East.

#### Instructions 

##### Step 1. Flash back to our Intro to SQL for Data Science course and begin by selecting all country codes in the Middle East as a single field result using SELECT, FROM, and WHERE.

In [20]:
pd.read_sql(
    """ 
        SELECT code
        FROM countries
        WHERE region = 'Middle East';
    """, con = engine)

code
0   ARE
1   ARM
2   AZE
3   BHR
4   GEO
5   IRQ
6   ISR
7   YEM
8   JOR
9   KWT
10  CYP
11  LBN
12  OMN
13  QAT
14  SAU
15  SYR
16  TUR
17  PSE

##### Step 2. 
Comment out the answer to the previous tab by surrounding it in `/*` and `*/`. You'll come back to it!

Below the commented code, select only unique languages by name appearing in the languages table.

Order the resulting single field table by name in ascending order.

In [21]:
pd.read_sql(
    """ /*
        SELECT code
        FROM countries
        WHERE region = 'Middle East';
        */

        SELECT DISTINCT name
        FROM languages
        ORDER BY name; 
    """, con = engine)

name
0           Afar
1      Afrikaans
2          Akyem
3       Albanian
4       Alsatian
..           ...
391      siSwati
392         sign
393       tribal
394      unknown
395  unspecified

[396 rows x 1 columns]

##### Step 3. Now combine the previous two queries into one query:

Add a **WHERE IN** statement to the **SELECT DISTINCT** query, and use the commented out query from the first instruction in there. That way, you can determine the `unique languages` spoken in the `Middle East`.

Carefully review this result and its code after completing it. It serves as a great example of subqueries, which are the focus of Chapter 4.

In [22]:
pd.read_sql(
    """ 
        SELECT DISTINCT name
        FROM languages
        WHERE code IN
          
          -- Subquery
          ( SELECT code
            FROM countries
            WHERE region = 'Middle East')
        ORDER BY name;
    """, con = engine)

name
0        Arabic
1       Aramaic
2      Armenian
3   Azerbaijani
4         Azeri
5       Baluchi
6     Bulgarian
7    Circassian
8       English
9         Farsi
10     Filipino
11       French
12     Georgian
13        Greek
14       Hebrew
15        Hindi
16       Indian
17      Kurdish
18        Other
19      Persian
20     Romanian
21      Russian
22       Syriac
23      Turkish
24      Turkmen
25         Urdu
26  unspecified

### Relating semi-join to a tweaked inner join

Let's revisit the code from the `previous exercise`, which retrieves `languages` spoken in the `Middle East` (line 21).

Sometimes **problems solved with semi-joins can also be solved using an inner join.**

In [23]:
pd.read_sql(
    """ 
        SELECT languages.name AS language
        FROM languages
        INNER JOIN countries
        ON languages.code = countries.code
        WHERE region = 'Middle East'
        ORDER BY language;    
    """, con = engine)

language
0        Arabic
1        Arabic
2        Arabic
3        Arabic
4        Arabic
..          ...
61      Turkmen
62         Urdu
63         Urdu
64         Urdu
65  unspecified

[66 rows x 1 columns]

This inner join isn't quite right. 

**What is missing from this second code block to get it to match with the correct answer produced by the first block?**

Tha answers is **There's no use on retrieving 'Arabic' multiple times; you only care about DISTINCT languages here.**

In [24]:
pd.read_sql(
    """ 
        SELECT DISTINCT languages.name AS language
        FROM languages
        INNER JOIN countries
        ON languages.code = countries.code
        WHERE region = 'Middle East'
        ORDER BY language;    
    """, con = engine)

language
0        Arabic
1       Aramaic
2      Armenian
3   Azerbaijani
4         Azeri
5       Baluchi
6     Bulgarian
7    Circassian
8       English
9         Farsi
10     Filipino
11       French
12     Georgian
13        Greek
14       Hebrew
15        Hindi
16       Indian
17      Kurdish
18        Other
19      Persian
20     Romanian
21      Russian
22       Syriac
23      Turkish
24      Turkmen
25         Urdu
26  unspecified

## 5. Diagnosing problems using anti-join

Another powerful join in `SQL` is the **`anti-join`**. It is particularly useful in identifying which records are causing an incorrect number of records to appear in join queries.

You will also see another example of a subquery here, as you saw in the first exercise on semi-joins. Your goal is to identify the currencies used in `Oceanian countries`!

#### Instructions 

**Step 1.** Begin by determining the number of countries in countries that are listed in `Oceania` using **SELECT, FROM**, and **WHERE**.

In [26]:
pd.read_sql(
    """
        SELECT c1.code, c1.country_name, c2.basic_unit AS currency
        FROM countries AS c1
        LEFT JOIN currencies AS c2
        ON c1.code = c2.code
        WHERE c1.continent = 'Oceania';
    """, con = engine)

code                     country_name                currency
0   ASM                   American Samoa                    None
1   AUS                        Australia       Australian dollar
2   FJI                     Fiji Islands                    None
3   GUM                             Guam                    None
4   KIR                         Kiribati       Australian dollar
5   MHL                 Marshall Islands    United States dollar
6   FSM  Micronesia, Federated States of                    None
7   NRU                            Nauru       Australian dollar
8   PLW                            Palau    United States dollar
9   PNG                 Papua New Guinea  Papua New Guinean kina
10  MNP         Northern Mariana Islands                    None
11  PYF                 French Polynesia               CFP franc
12  SLB                  Solomon Islands  Solomon Islands dollar
13  WSM                            Samoa             Samoan tala
14  TON                            Tonga          Tongan paʻanga
15  TUV                           Tuvalu       Australian dollar
16  TUV                           Tuvalu         Tuvaluan dollar
17  NCL                    New Caledonia               CFP franc
18  NZL                      New Zealand      New Zealand dollar
19  VUT                          Vanuatu            Vanuatu vatu

**Step 3.** Note that not all countries in `Oceania` were listed in the resulting inner join with currencies. Use an anti-join to determine which countries were not included!

Use **NOT IN** and (**SELECT** `code` **FROM** `currencies`) as a **`subquery`** to get the `country code` and `country name` for the `Oceanian countries` that are not included in the `currencies` table.

In [28]:
pd.read_sql(
    """
        SELECT code, country_name
        FROM countries
        WHERE continent = 'Oceania'
        AND code NOT IN
        -- 2. Subquery
          (SELECT code
           FROM currencies);
    """, con = engine)

code                     country_name
0  ASM                   American Samoa
1  FJI                     Fiji Islands
2  GUM                             Guam
3  FSM  Micronesia, Federated States of
4  MNP         Northern Mariana Islands

### Set theory challenge

Your task here will be to incorporate two of **UNION/UNION ALL/INTERSECT/EXCEPT** to solve a challenge involving `three tables`.

In addition, you will use a `subquery` as you have in the `last two exercises`! This will be great practice as you hop into subqueries more in the next chapter.

#### Instructions

Identify the country codes that are included in either economies or currencies but not in populations.

Use that result to determine the names of cities in the countries that match the specification in the previous instruction

In [29]:
pd.read_sql(
    """
        SELECT name
        FROM cities AS c1
        -- Choose only records matching the result of multiple set theory clauses
        WHERE country_code IN
        (
            -- Select appropriate field from economies AS e
            SELECT e.code
            FROM economies AS e
            -- Get all additional (unique) values of the field from currencies AS c2  
            UNION
            SELECT c2.code
            FROM currencies AS c2
            -- Exclude those appearing in populations AS p
            EXCEPT
            SELECT p.country_code
            FROM populations AS p
        );
    """, con = engine)

name
0        Bucharest
1        Kaohsiung
2  New Taipei City
3         Taichung
4           Tainan
5           Taipei